File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < inorder.length; i ++) {
            map.put(inorder[i], i);
        }
        return buildTreeHelper(preorder, 0, 0, inorder.length - 1, map);
    }
    private TreeNode buildTreeHelper(int[] preorder, int start, int left, int right, Map<Integer, Integer> map) {
        TreeNode tree = new TreeNode();
        if (left > right) {
            return null;
        }
        if (right >= preorder.length) {
            return null;
        }
        tree.val = preorder[start];
        int mid = map.get(tree.val);
        tree.left = buildTreeHelper(preorder, start + 1, left, mid - 1, map);
        tree.right = buildTreeHelper(preorder, start + (mid - left) + 1, mid + 1, right, map);
        return tree;
    }  
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
      def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def kthSmallestHelper(root: Optional[TreeNode], nums : List[int]) -> None:
            if root is None:
                return 
            kthSmallestHelper(root.left, nums)
            nums.append(root.val)
            kthSmallestHelper(root.right, nums)
        nums = []
        kthSmallestHelper(root, nums)
        return nums[k - 1] 
    
        